keras_contrib/utils/kito.py

```diff
@@ -0,0 +1,309 @@
+"""
+Reduce neural net structure (Conv + BN -> Conv)
+Also works:
+DepthwiseConv2D + BN -> DepthwiseConv2D
+SeparableConv2D + BN -> SeparableConv2D
+
+This code takes on input trained Keras model and optimize layer structure and weights in such a way
+that model became much faster (~30%), but works identically to initial model. It can be extremely
+useful in case you need to process large amount of images with trained model. Reduce operation was
+tested on all Keras models zoo. See comparison table and full description by link:
+https://github.com/ZFTurbo/Keras-inference-time-optimizer
+Author: Roman Solovyev (ZFTurbo)
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+
+
+def get_keras_sub_version():
+    from keras import __version__
+    type = int(__version__.split('.')[1])
+    return type
+
+
+def get_input_layers_ids(model, layer, verbose=False):
+    res = dict()
+    for i, l in enumerate(model.layers):
+        layer_id = str(id(l))
+        res[layer_id] = i
+
+    inbound_layers = []
+    layer_id = str(id(layer))
+    for i, node in enumerate(layer._inbound_nodes):
+        node_key = layer.name + '_ib-' + str(i)
+        if get_keras_sub_version() == 1:
+            network_nodes = model._container_nodes
+        else:
+            network_nodes = model._network_nodes
+        if node_key in network_nodes:
+            for inbound_layer in node.inbound_layers:
+                inbound_layer_id = str(id(inbound_layer))
+                inbound_layers.append(res[inbound_layer_id])
+    return inbound_layers
+
+
+def get_output_layers_ids(model, layer, verbose=False):
+    res = dict()
+    for i, l in enumerate(model.layers):
+        layer_id = str(id(l))
+        res[layer_id] = i
+
+    outbound_layers = []
+    layer_id = str(id(layer))
+    for i, node in enumerate(layer._outbound_nodes):
+        node_key = layer.name + '_ib-' + str(i)
+        if get_keras_sub_version() == 1:
+            network_nodes = model._container_nodes
+        else:
+            network_nodes = model._network_nodes
+        if node_key in network_nodes:
+            outbound_layer_id = str(id(node.outbound_layer))
+            if outbound_layer_id in res:
+                outbound_layers.append(res[outbound_layer_id])
+            else:
+                print('Warning, some problem with outbound node on layer {}!'.format(layer.name))
+    return outbound_layers
+
+
+def get_copy_of_layer(layer, verbose=False):

```

Comment:
how do users do this for their own non-standard layers?

Reply:
They need to change the code ( Like here:

    # DeepLabV3+ non-standard layer
    if layer.__class__.__name__ == 'BilinearUpsampling':
        from neural_nets.deeplab_v3_plus_model import BilinearUpsampling

I'm not sure how to make it universal, since parameters are different.

Reply:
You can also have a look at keras itself registers functions. If you look at the python file for loss, at the very bottom there is some registration code.